product_cost_price_avco_sync/models/stock_move.py

```diff
@@ -0,0 +1,101 @@
+# Copyright 2019 Carlos Dauden - Tecnativa <carlos.dauden@tecnativa.com>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from datetime import timedelta
+
+from odoo import fields, models
+from odoo.tools.float_utils import float_compare
+
+
+class StockMove(models.Model):
+    _inherit = 'stock.move'
+
+    def write(self, vals):
+        """ """
+        res = super(StockMove, self).write(vals)
+        if ('price_unit' in vals and
+                not self.env.context.get('skip_avco_sync')):
+            self.cost_price_avco_sync()
+        return res
+
+    def _create_price_history(self, value):
+        PriceHistory = self.env['product.price.history']
+        for move in self:
+            PriceHistory.create({
+                'product_id': move.product_id.id,
+                'cost': value,
+                'datetime': move.date,
+                'company_id': move.company_id.id,
+            })
+
+    def _previous_instant_date(self):
+        """ Returns previous instant before move was done"""
+        self.ensure_one()
+        date_with_delta = fields.Datetime.from_string(
+            self.date) - timedelta(seconds=1)
+        return fields.Datetime.to_string(date_with_delta)
+
+    def _remove_after_history_price(self):
+        PriceHistory = self.env['product.price.history']
+        for move in self:
+            PriceHistory.search([
+                ('company_id', '=', move.company_id.id),
+                ('product_id', '=', move.product_id.id),
+                ('datetime', '>=', move.date),
+            ]).unlink()
+
+    def get_average_price(self, previous_qty, previous_price):
+        new_std_price = ((previous_qty * previous_price) + (
+            self.price_unit * self.quantity_done)) / (previous_qty +
+                                                      self.quantity_done)
+        return new_std_price
+
+    def cost_price_avco_sync(self):
+        procesed_moves = set()
+        for move in self.sorted('date'):
+            if (move.id in procesed_moves or move.state != 'done' or
+                    not move._is_in() or
+                    move.product_id.cost_method != 'average'):
+                continue
+            previous_instant = move._previous_instant_date()
+            previous_qty = move.product_id.with_context(
+                to_date=previous_instant
+            ).qty_available
+            previous_price = move.product_id.get_history_price(
+                move.company_id.id, previous_instant) if previous_qty else 0.0
+            price = move.get_average_price(previous_qty, previous_price)
+            move._remove_after_history_price()
+            move._create_price_history(price)
```

Comment:
There's duplicated code that can be avoided reordering a bit the code and adding an extra `if`, but it's worth for avoiding duplicating logic. We can see it together if you want.

Reply:
I do not know if I understand well the alternative that you propose, I have considered several options but the code is not so understandable.
Feel free to make a suggestion or a commit.